File: src/python/day05.py
# Input:
example_input = """
3-5
10-14
16-20
12-18

1
5
8
11
17
32
"""


def parse_input(input: str) -> tuple[list[tuple[int, int]], list[int]]:
    res = input.strip().split("\n\n")
    fresh_ranges_str, available_ids_str = res[0], res[1]
    fresh_ranges = []
    for line in fresh_ranges_str.split("\n"):
        start, end = map(int, line.split("-"))
        fresh_ranges.append((start, end))
    available_ids = [int(line) for line in available_ids_str.split("\n")]
    return fresh_ranges, available_ids


def how_many_fresh_check(
    fresh_ranges: list[tuple[int, int]], available_ids: list[int]
) -> int:
    fresh_count = 0
    for ingredient_id in available_ids:
        for start, end in fresh_ranges:
            if start <= ingredient_id <= end:
                fresh_count += 1
                break
    return fresh_count


def how_many_fresh_all(fresh_ranges: list[tuple[int, int]]) -> int:
    fresh_ranges.sort()
    res = [fresh_ranges[0]]
    for start, end in fresh_ranges:
        _, last_end = res[-1]
        if last_end >= start - 1:
            last_start, last_end = res.pop()
            res.append((min(last_start, start), max(last_end, end)))
        else:
            res.append((start, end))

    return sum([end - start + 1 for start, end in res])


input = open("input/day05.txt", "r").read()
# input = example_input.strip()

fresh_ranges, available_ids = parse_input(input)
print(how_many_fresh_check(fresh_ranges, available_ids))
print(how_many_fresh_all(fresh_ranges))

File: src/rust/day05.rs
use std::cmp::Ordering;
use std::ops::Range;

use itertools::Itertools;

type Ingredient = usize;
type IngredientRange = Range<Ingredient>;
type IngredientRanges = Vec<IngredientRange>;

fn parse_input(input: &str) -> (IngredientRanges, Vec<Ingredient>) {
    let (ingredient_ranges, ingredient_list) = input.split_once("\n\n").unwrap();
    let ingredient_ranges = ingredient_ranges
        .lines()
        .map(|line| {
            let (start_str, end_str) = line.split_once('-').unwrap();
            let start = start_str.parse::<Ingredient>().unwrap();
            let end = end_str.parse::<Ingredient>().unwrap() + 1;
            start..end
        })
        .collect();
    let ingredient_list = ingredient_list
        .lines()
        .map(|line| line.parse::<Ingredient>().unwrap())
        .collect();
    (ingredient_ranges, ingredient_list)
}

fn merge_ranges(ranges: IngredientRanges) -> IngredientRanges {
    ranges
        .into_iter()
        .sorted_by_key(|range| range.start)
        .coalesce(|prev, next| {
            if prev.end >= next.start {
                Ok(prev.start..prev.end.max(next.end))
            } else {
                Err((prev, next))
            }
        })
        .collect()
}

fn is_in_range(range: &IngredientRange, ingredient: Ingredient) -> Ordering {
    if range.contains(&ingredient) {
        Ordering::Equal
    } else if range.start > ingredient {
        Ordering::Greater
    } else {
        Ordering::Less
    }
}

pub fn how_many_fresh_check(input: &str) -> usize {
    let (ingredient_ranges, ingredient_list) = parse_input(input);
    let merged_ranges = merge_ranges(ingredient_ranges);
    ingredient_list
        .into_iter()
        .filter(|&ingredient| {
            merged_ranges
                .binary_search_by(|range| is_in_range(range, ingredient))
                .is_ok()
        })
        .count()
}

pub fn how_many_fresh_all(input: &str) -> usize {
    let (ingredient_ranges, _) = parse_input(input);
    merge_ranges(ingredient_ranges)
        .into_iter()
        .map(|range| range.len())
        .sum()
}

pub fn main() {
    let input = std::fs::read_to_string("input/day05.txt").expect("Input file not found");
    let now = std::time::Instant::now();
    println!("PART 1 = {}", how_many_fresh_check(&input));
    println!("PART 2 = {}", how_many_fresh_all(&input));
    println!("Execution time: {:?}", now.elapsed());
}
